### DISCUSSION TITLE
How can I use a single table definition (entity?) with multiple schemas?

### DISCUSSION THREAD

**Discussion by tukusejssirs** [mikro-orm/mikro-orm#2439] [Q&A] [answered]
Note: I am very new to TypeScript and ORM, so bear with me. :wink:

---

I use PostgreSQL and in my use case I need to define a set of tables per schema. I create a schema per machine/device I monitor.

Now, I understand how I can create an entity using `@Entity()` decorator. I’ve also read [Using Multiple Schemas](https://mikro-orm.io/docs/multiple-schemas/), but it directly binds the entity with a specific `schema.table`.

In my use case, I’d like to create a single entity per table, but use it with that table in any schema (e.g. I’d like to use `table` entity with `m1.table`, `m2.table`, …, which all have the same columns, types, etc). I’d like to change the schema dynamically at runtime.

Is it possible with Mikro-ORM?

Thanks for the answer! :pray:

**Comment by B4nan** [comment, marked as answer]
v5 supports this already via wildcard schema:

https://mikro-orm.io/docs/next/multiple-schemas/#wildcard-schema

**Reply by tukusejssirs** [reply]
Thanks, @B4nan, for such a quick answer!

Anyway, how stable is v5? As I understand it, it is still in development. Is there any ETA on its release date?

I didn’t mention it, but I use Nest.js and it seems like there is no `@micro-orm/nestjs@5` package yet (latest stable is v4.3.1 and next 4.3.0-dev.1). I have no idea if `@micro-orm/nestjs@4.3.1` would work with `@micro-orm/{core,postgresql}@5`.

**Reply by B4nan** [reply]
Yes it does support v5, it is almost complete, planning few more changes, but if it will be something semi-breaking, it will be configurable. 

https://github.com/mikro-orm/nestjs/blob/master/package.json#L44

**Reply by tukusejssirs** [reply]
Well, I didn’t have luck installing v5 in a Nest.js project:

<details>

   <summary><code>npm i @mikro-orm/core@5 @mikro-orm/postgresql@5</code></summary>

```js
npm WARN ERESOLVE overriding peer dependency
npm WARN While resolving: controlserver@0.0.1
npm WARN Found: @mikro-orm/core@4.5.9
npm WARN node_modules/@mikro-orm/core
npm WARN   peer @mikro-orm/core@"^4.0.0" from @mikro-orm/knex@4.5.9
npm WARN   node_modules/@mikro-orm/knex
npm WARN     @mikro-orm/knex@"^4.5.9" from @mikro-orm/postgresql@4.5.9
npm WARN     node_modules/@mikro-orm/postgresql
npm WARN   3 more (@mikro-orm/nestjs, @mikro-orm/postgresql, the root project)
npm WARN 
npm WARN Could not resolve dependency:
npm WARN peer @mikro-orm/core@"^4.0.0||^5.0.0-dev.0" from @mikro-orm/nestjs@4.3.1
npm WARN node_modules/@mikro-orm/nestjs
npm WARN   @mikro-orm/nestjs@"^4.3.1" from the root project
npm WARN ERESOLVE overriding peer dependency
npm WARN While resolving: controlserver@0.0.1
npm WARN Found: @mikro-orm/core@4.5.9
npm WARN node_modules/@mikro-orm/core
npm WARN   peer @mikro-orm/core@"^4.0.0" from @mikro-orm/knex@4.5.9
npm WARN   node_modules/@mikro-orm/knex
npm WARN     @mikro-orm/knex@"^4.5.9" from @mikro-orm/postgresql@4.5.9
npm WARN     node_modules/@mikro-orm/postgresql
npm WARN   3 more (@mikro-orm/nestjs, @mikro-orm/postgresql, the root project)
npm WARN 
npm WARN Could not resolve dependency:
npm WARN peer @mikro-orm/core@"^4.0.0" from @mikro-orm/postgresql@4.5.9
npm WARN node_modules/@mikro-orm/postgresql
npm WARN   peerOptional @mikro-orm/postgresql@"^4.0.0" from @mikro-orm/core@4.5.9
npm WARN   1 more (the root project)
npm ERR! code ETARGET
npm ERR! notarget No matching version found for @mikro-orm/core@5.
npm ERR! notarget In most cases you or one of your dependencies are requesting
```

</details>

**Reply by B4nan** [reply]
you need to install from `next` tag, what you tried is installing from `latest`

**Reply by tukusejssirs** [reply]
Still get an error:

<details>

   <summary><code>npm i @mikro-orm/core@next @mikro-orm/postgresql@next @mikro-orm/nestjs@next</code></summary>

```js
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! While resolving: controlserver@0.0.1
npm ERR! Found: @mikro-orm/core@5.0.0-dev.501
npm ERR! node_modules/@mikro-orm/core
npm ERR!   @mikro-orm/core@"5.0.0-dev.501" from the root project
npm ERR!   peer @mikro-orm/core@"^4.0.0||^5.0.0-dev.0" from @mikro-orm/postgresql@5.0.0-dev.501
npm ERR!   node_modules/@mikro-orm/postgresql
npm ERR!     @mikro-orm/postgresql@"5.0.0-dev.501" from the root project
npm ERR!     peerOptional @mikro-orm/postgresql@"^4.0.0||^5.0.0-dev.0" from @mikro-orm/core@5.0.0-dev.501
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer @mikro-orm/core@"^4.5.7" from @mikro-orm/nestjs@4.3.0-dev.1
npm ERR! node_modules/@mikro-orm/nestjs
npm ERR!   @mikro-orm/nestjs@"4.3.0-dev.1" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR! 
npm ERR! See /home/ts/.npm/eresolve-report.txt for a full report.
```

</details>

**Reply by B4nan** [reply]
no not use `@mikro-orm/nestjs@next`, there are no automatic dev versions, so the one under next tag is outdate. use latest version there instead, so 4.3.1. (just drop the `@next`)

**Reply by B4nan** [reply]
https://github.com/mikro-orm/nestjs-realworld-example-app/tree/v5

**Reply by tukusejssirs** [reply]
@B4nan, I tried to install `@mikro-orm/sql-highlighter` with v5, but there is no `next` version of it (`latest` being `1.0.1`). Is it now incorporated into another package or it is simply not updated, yet?

**Reply by B4nan** [reply]
nope, only packages that are part of the main repository are v5. use latest from other packages like the highlighters or the nestjs adapter. there is no point in updating the highlighters, they work just fine
